src/v/cluster/topics_frontend.cc

```diff
@@ -305,29 +306,75 @@ make_error_result(const model::topic_namespace& tp_ns, std::error_code ec) {
     return topic_result(tp_ns, errc::topic_operation_error);
 }
 
-allocation_request make_allocation_request(const topic_configuration& cfg) {
+allocation_request
+make_allocation_request(const custom_assignable_topic_configuration& ca_cfg) {
+    // no custom assignments, lets allocator decide based on partition count
     allocation_request req;
-    req.partitions.reserve(cfg.partition_count);
-    for (auto p = 0; p < cfg.partition_count; ++p) {
-        req.partitions.emplace_back(
-          model::partition_id(p), cfg.replication_factor);
+    if (!ca_cfg.has_custom_assignment()) {
+        req.partitions.reserve(ca_cfg.cfg.partition_count);
+        for (auto p = 0; p < ca_cfg.cfg.partition_count; ++p) {
+            req.partitions.emplace_back(
+              model::partition_id(p), ca_cfg.cfg.replication_factor);
+        }
+    } else {
+        req.partitions.reserve(ca_cfg.custom_assignments.size());
+        for (auto& cas : ca_cfg.custom_assignments) {
+            allocation_constraints constraints;
+            constraints.hard_constraints.push_back(
+              ss::make_lw_shared<hard_constraint_evaluator>(
+                on_nodes(cas.replicas)));
+
+            req.partitions.emplace_back(
+              cas.id, cas.replicas.size(), std::move(constraints));
+        }
     }
     return req;
 }
 
+errc topics_frontend::validate_topic_configuration(
+  const custom_assignable_topic_configuration& assignable_config) {
+    if (!validate_topic_name(assignable_config.cfg.tp_ns)) {
+        return errc::invalid_topic_name;
+    }
+
+    if (assignable_config.cfg.partition_count < 1) {
+        return errc::topic_invalid_partitions;
+    }
+
+    if (assignable_config.cfg.replication_factor < 1) {
+        return errc::topic_invalid_replication_factor;
+    }
+
+    if (assignable_config.has_custom_assignment()) {
+        for (auto& custom : assignable_config.custom_assignments) {
+            if (
+              static_cast<int16_t>(custom.replicas.size())
+              != assignable_config.cfg.replication_factor) {
+                return errc::topic_invalid_replication_factor;
+            }
+        }
+    }
```

Comment:
do we cover validation where custom assignment does use the same node?

Reply:
yes, there is even a test for this, partition allocation will fail